RD-2600 & RD-2620 Improve handling component nodes in Topology widget

qooban opened the pull request:
### Changes

See commits list + comments inside the code for description about the changes.
My general conclusions are that this widget is another candidate for major refactoring. I will probably create a similar ticket to RD-2591 which will describe my findings.


### System tests

The following tests were done:
1. [Stage-UI-System-Test/775](https://jenkins.cloudify.co/view/UI%20Health%20View/job/Stage-UI-System-Test/775) (only for `topology_spec.ts`) - FAILED (after that I added https://github.com/cloudify-cosmo/cloudify-stage/pull/1433/commits/486d02e2058011baf96258488456de349c704288)
2. [Stage-UI-System-Test/777](https://jenkins.cloudify.co/blue/organizations/jenkins/Stage-UI-System-Test/detail/Stage-UI-System-Test/777/pipeline) (only for `topology_spec.ts`) - PASSED
3. [Stage-UI-System-Test/778](https://jenkins.cloudify.co/blue/organizations/jenkins/Stage-UI-System-Test/detail/Stage-UI-System-Test/778/pipeline) (only for `topology_spec.ts`) - PASSED
4. [Stage-UI-System-Test/783](https://jenkins.cloudify.co/blue/organizations/jenkins/Stage-UI-System-Test/detail/Stage-UI-System-Test/783/pipeline/) (full set of tests) - PASSED

qooban commented on widgets/topology/src/Topology.tsx:
```diff
@@ -205,7 +205,9 @@ export default class Topology extends React.Component {
             [...deploymentsToExpand].forEach(deploymentId => {
                 const expandedNodeData = findExpandedNode(topology, deploymentId);
-                if (!expandedNodeData) {
+                const componentDeploymentsData = data.componentDeploymentsData[deploymentId];
+
+                if (!expandedNodeData || !componentDeploymentsData) {
```
That is one of the main fixes in this PR.
Before expanded topology was created (L214) despite no `componentDeploymentsData` was present.

qooban commented on widgets/topology/src/widget.tsx:
```diff
@@ -145,10 +147,19 @@ Stage.defineWidget({
             },
             icons
-        } = _.isEmpty(data) ? { rawBlueprintData: { data: { id: '' }, layout: {} } } : data;
+        } = _.isEmpty(data)
+            ? {
+                  blueprintDeploymentData: {},
+                  componentDeploymentsData: {},
+                  rawBlueprintData: { data: { id: '' }, layout: {} },
+                  icons: {}
+              }
+            : data;
```
This is the second fix - `componentDeploymentsData` was not initialized properly in case of empty `data`.

qooban commented on widgets/topology/src/widget.tsx:
```diff
@@ -145,10 +147,19 @@ Stage.defineWidget({
+        const blueprintId = deploymentId ? id : castArray(toolbox.getContext().getValue('blueprintId'))[0];
 
-        const deploymentId = toolbox.getContext().getValue('deploymentId');
-        const blueprintId = deploymentId ? id : toolbox.getContext().getValue('blueprintId');
```
Another fix is to use only the first resource ID.

qooban commented on test/cypress/integration/widgets/topology_spec.ts:
```diff
@@ -1,15 +1,22 @@
     const resourcePrefix = 'topology_test_';
     const getNodeTopologyButton = index => cy.get(`.nodeTopologyButton:eq(${index})`);
-    const waitForDeploymentToBeInstalled = deploymentId =>
-        waitUntilEmpty(`deployments?id=${deploymentId}&deployment_status=in_progress`);
+
+    const waitForDeploymentToBeInstalled = deploymentId => {
+        cy.log(`Waiting for deployment ${deploymentId} to be installed.`);
+        const startedInstallWorkflowsOnDeployment = `executions?_include=id,workflow_id,status&deployment_id=${deploymentId}&workflow_id=install&status=started`;
+
+        waitUntilNotEmpty(startedInstallWorkflowsOnDeployment);
+        waitUntilEmpty(startedInstallWorkflowsOnDeployment);
+    };
```
Previous way for waiting for deployment to be installed had some flaws, so I improved it a bit by first waiting for execution to be in `started` state, then wait for it to be finished (be not in `started` state anymore).

qooban commented on test/cypress/integration/widgets/topology_spec.ts:
```diff
@@ -118,6 +125,20 @@ describe('Topology', () => {
+            cy.wait('@fetchLayout', { requestTimeout: minutesToMs(pollingTime + 1) });
+            cy.waitUntilPageLoaded();
+
+            cy.log('Waiting until animation in topology canvas is finished');
+            // eslint-disable-next-line cypress/no-unnecessary-waiting
+            cy.wait(2000);
```
Despite that I know that Cypress is waiting for animation to finish I didn't find better way to do that. 
I tried to wait for different elements, modifing Cypress config (changing value of `animationDistanceThreshold`), but without success.

Gelio commented on widgets/topology/src/Topology.tsx:
```diff
@@ -205,7 +205,9 @@ export default class Topology extends React.Component {
             [...deploymentsToExpand].forEach(deploymentId => {
                 const expandedNodeData = findExpandedNode(topology, deploymentId);
-                if (!expandedNodeData) {
+                const componentDeploymentsData = data.componentDeploymentsData[deploymentId];
```
Minor: we probably could use this variable instead of accessing it again on L214 :slightly_smiling_face:

qooban replied:
Right, of course, good catch. Reused in L214.

Gelio commented on widgets/topology/src/widget.tsx:
```diff
@@ -145,7 +145,14 @@ Stage.defineWidget({
-        } = _.isEmpty(data) ? { rawBlueprintData: { data: { id: '' }, layout: {} } } : data;
+        } = _.isEmpty(data)
+            ? {
+                  blueprintDeploymentData: {},
```
If I'm not mistaken, the destructing above takes `processedBlueprintData` from `data` and extracts it as a variable `blueprintDeploymentData`. Thus, should it not be named `processedBlueprintData`?

```suggestion
                  processedBlueprintData: {},
```

qooban replied:
Good catch. Suggestion applied.

Gelio commented on test/cypress/support/resource_commons.ts:
```diff
@@ -1,19 +1,28 @@
-// @ts-nocheck File not migrated fully to TS
```
Yay :smile:

Gelio commented on test/cypress/support/resource_commons.ts:
```diff
@@ -1,19 +1,28 @@
+function waitUntil(
+    resource: string,
+    predicate: (response: Cypress.Response) => boolean,
+    { search, numberOfRetriesLeft, waitingInterval } = {
+        search: '',
+        numberOfRetriesLeft: 60,
+        waitingInterval: 1000
+    }
```
Minor:

We probably want to apply the default values to the destructured parameters:

```ts
{ search = '', numberOfRetriesLeft = 60, waitingInterval = 1000 } = {}
```

qooban replied:
Yeah, right, changed.

Gelio commented on test/cypress/support/resource_commons.ts:
```diff
@@ -22,6 +31,14 @@ export function waitUntilEmpty(resource, search, numberOfRetriesLeft = 60, waiti
     });
 }
 
+export function waitUntilNotEmpty(resource: string, search: string, numberOfRetriesLeft = 60, waitingInterval = 1000) {
+    waitUntil(resource, response => !_.isEmpty(response.body.items), { search, numberOfRetriesLeft, waitingInterval });
+}
+
+export function waitUntilEmpty(resource: string, search: string, numberOfRetriesLeft = 60, waitingInterval = 1000) {
```
We probably don't want to repeat the default values for parameter twice - once in those special functions, and another time in the base function :smile: Those can be safely removed after implementing the comment I left on the default values in `waitUntil` :slightly_smiling_face:

qooban replied:
That's true. 
Also, I refactored it a bit to have a bit better signature (had to update 3 other files, just 3 lines in total).

Gelio replied:
Looks good, thanks :+1:

Gelio commented on test/cypress/support/resource_commons.ts:
```diff
@@ -1,19 +1,28 @@
 }
 
-export function waitUntilEmpty(resource, search, numberOfRetriesLeft = 60, waitingInterval = 1000) {
+function waitUntil(
```
I like the pattern of having a base function with dependency injection :+1: Cool idea

qooban replied:
Thanks 👍🏻 🙂

Gelio commented on test/cypress/integration/widgets/topology_spec.ts:
```diff
@@ -1,12 +1,18 @@
-    const waitForDeploymentToBeInstalled = deploymentId =>
-        waitUntilEmpty(`deployments?id=${deploymentId}&deployment_status=in_progress`);
+
+    const waitForDeploymentToBeInstalled = deploymentId => {
```
Minor, very optional:

Adding a type annotation to the parameter would make it easier to migrate the file in the future :slightly_smiling_face:

qooban replied:
Good point. Added.

Tried to convert it fully, but it'll need a bit more attention (e.g. `test/cypress/support/executions.ts` refactoring), so I think it's not for this PR.

Gelio commented on test/cypress/integration/widgets/topology_spec.ts:
```diff
@@ -1,12 +1,18 @@
+        cy.log(`Waiting for deployment ${deploymentId} to be installed.`);
+        const startedInstallWorkflowsOnDeployment = `executions?_include=id,workflow_id,status&deployment_id=${deploymentId}&workflow_id=install&status=started`;
+
+        waitUntilNotEmpty(startedInstallWorkflowsOnDeployment);
```
Looking at it, this may look like a mistake and could be removed during some future refactor without any errors immediately surfacing. How about adding a comment explaining why we need first to wait for the execution to start and that this is deliberate?

I believe this will improve the maintainability of the code :slightly_smiling_face:

qooban replied:
Sure, added a comment.

Gelio commented on widgets/topology/src/widget.tsx:
```diff
@@ -40,8 +41,9 @@ Stage.defineWidget({
-        const deploymentId = toolbox.getContext().getValue('deploymentId');
-        const blueprintId = toolbox.getContext().getValue('blueprintId');
+        // TODO(RD-2130): Use common utility function to get only the first ID
+        const deploymentId = castArray(toolbox.getContext().getValue('deploymentId'))[0];
```
Good fixes :+1:

qooban replied:
Thanks 👍🏻 🙂 

Of course the best would be to have RD-2130 implemented and just reuse common functions, but still - we're a bit better right now.

Gelio commented on test/cypress/integration/widgets/topology_spec.ts:
```diff
@@ -1,8 +1,9 @@
+import { minutesToMs, waitUntilEmpty, waitUntilNotEmpty } from '../../support/resource_commons';
 
 describe('Topology', () => {
+    const pollingTime = 5;
```
Minor:

Can we add a comment saying this is in seconds?

qooban replied:
Sure, added.

Gelio commented on test/cypress/integration/widgets/topology_spec.ts:
```diff
@@ -125,13 +126,18 @@ describe('Topology', () => {
             cy.log('Waiting for topology data to be fetched');
-            cy.get('.widgetLoader.active').should('be.visible');
-            cy.get('.widgetLoader.active').should('not.exist');
+            cy.wait('@fetchLayout', { requestTimeout: minutesToMs(pollingTime + 1) });
```
Do we really want to wait 6 minutes for this request :thinking: ?

qooban replied:
Of course not, my mistake, so thanks for good catch! 🙂 The intention was to wait `pollingTime` [seconds] + 1 [second].

Now I think it would be more reliable to wait 2 * `pollingTime` [seconds], so:
1. I have added another helper function `secondsToMs`
2. Changed `requestTimeout` to be 2 * `pollingTime`

Gelio replied:
Awesome :smile: Looks better now :smile:

Gelio reviewed (approved):
Great PR, clear separation of commits made it easy to review and the comments also helped :+1: 

I did not run it myself in the UI, but looking at the tests we should be good :slightly_smiling_face: 

I have added some comments to potentially improve the quality :smile:

qooban commented:
Thanks for the reviews. I think I addressed all the comments.

Scheduled another full system test execution: [Stage-UI-System-Test/785](https://jenkins.cloudify.co/blue/organizations/jenkins/Stage-UI-System-Test/detail/Stage-UI-System-Test/785/pipeline/)

kubama commented on test/cypress/integration/widgets/topology_spec.ts:
```diff
@@ -1,18 +1,20 @@
 
 describe('Topology', () => {
-    const pollingTime = 5;
+    const pollingTime = 5; // seconds
```
How about adding the `seconds` phrase to the variable name?

qooban replied:
Yep, also good. Changed.

qooban commented:
All tests passed in [Stage-UI-System-Test/785](https://jenkins.cloudify.co/blue/organizations/jenkins/Stage-UI-System-Test/detail/Stage-UI-System-Test/785/pipeline/), so I'll just wait for the final review from @Gelio .

Gelio reviewed (approved):
LGTM :+1: Thanks for addressing all the comments

May the topology spec finally be reliable :smile:
